safehttp/task.go

```diff
@@ -0,0 +1,256 @@
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	t.commitPhase(NoContentResponse{})
+	t.rw.WriteHeader(int(StatusNoContent))
+	return Result{}
+}
+
+// WriteError writes an error response (400-599) according to the provided
+// status code.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) WriteError(code StatusCode) Result {
+	// TODO: accept custom error responses that need to go through the dispatcher.
+	if t.writtenError {
+		panic("ResponseWriter.WriteError called twice")
+	}
+	t.written = true
+	t.writtenError = true
+	resp := &ErrorResponse{Code: code}
+	t.errorPhase(resp)
+	http.Error(t.rw, http.StatusText(int(resp.Code)), int(resp.Code))
+	return Result{}
+}
+
+// Redirect responds with a redirect to the given url, using code as the status code.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) Redirect(r *IncomingRequest, url string, code StatusCode) Result {
+	if code < 300 || code >= 400 {
+		panic("wrong method called")
+	}
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	http.Redirect(t.rw, r.req, url, int(code))
+	return Result{}
+}
+
+// Header returns the collection of headers that will be set
+// on the response. Headers must be set before writing a
+// response (e.g. Write, WriteTemplate).
+func (t *Task) Header() Header {
+	return t.header
+}
+
+// SetCookie adds a Set-Cookie header to the provided ResponseWriter's headers.
+// The provided cookie must have a valid Name, otherwise an error will be
+// returned.
+func (t *Task) SetCookie(c *Cookie) error {
+	return t.header.addCookie(c)
+}
+
+// SetCode allows setting a response status. If the response was already
+// written, trying to set the status code will have no effect. This method will
+// panic if an invalid status code is passed (i.e. not in the range 1XX-5XX).
+//
+// If SetCode was called before NoContent, Redirect or WriteError, the status
+// code set by the latter will be the actual response status.
+//
+// TODO(empijei@, kele@, maramihali@): decide what should be done if the
+// code passed is either 3XX (redirect) or 4XX-5XX (client/server error).
+func (t *Task) SetCode(code StatusCode) {
+	if t.written {
+		return
+	}
+	if code < 100 || code >= 600 {
+		panic("invalid status code")
+	}
+	t.code = code
+}
+
+// commitPhase calls the Commit phases of all the interceptors. This stage will
+// run before a response is written to the ResponseWriter. If a response is
+// written to the ResponseWriter in a Commit phase then the Commit phases of the
+// remaining interceptors won't execute.
+//
+// TODO: BIG WARNING, if ResponseWriter.Write and ResponseWriter.WriteTemplate
+// are called in Commit then this will recurse. CommitResponseWriter was an
+// attempt to prevent this by not giving access to Write and WriteTemplate in
+// the Commit phase.
```

Comment:
```suggestion
```

Reply:
Yeah, I haven't changed the comments yet at all.